src/core/net/ip6.cpp

```diff
@@ -603,27 +622,240 @@ otError Ip6::HandleOptions(Message &aMessage, Header &aHeader, bool &aForward)
+    uint16_t       payloadFragment    = 0;
+    uint16_t       offset             = 0;
+    uint32_t       maxPayloadFragment = (1028 - aMessage.GetOffset() - sizeof(fragmentHeader)) & 0xfffffff8;
+    uint32_t       payloadLeft        = aMessage.GetLength() - aMessage.GetOffset();
+
+    if (aMessage.GetLength() > kMaxAssembledDatagramLength)
+    {
+        ExitNow(error = OT_ERROR_NO_BUFS);
+    }
+
+    VerifyOrExit(aMessage.Read(0, sizeof(header), &header) == sizeof(header), error = OT_ERROR_PARSE);
+    header.SetNextHeader(kProtoFragment);
+
+    fragmentHeader.Init();
+    fragmentHeader.SetIdentification(Random::NonCrypto::GetUint32());
+    fragmentHeader.SetNextHeader(aIpProto);
+    fragmentHeader.SetMoreFlag();
+
+    while (payloadLeft != 0)
+    {
+        if (payloadLeft < maxPayloadFragment)
+        {
+            fragmentHeader.ClearMoreFlag();
+            payloadFragment = payloadLeft;
+            payloadLeft     = 0;
+            otLogDebgIp6("Last Fragment");
+        }
+        else
+        {
+            payloadLeft -= maxPayloadFragment;
+            payloadFragment = maxPayloadFragment;
+        }
+        offset = fragmentCnt * (maxPayloadFragment >> 3);
+        fragmentHeader.SetOffset(offset);
+        VerifyOrExit((fragment = NewMessage(0)) != NULL, error = OT_ERROR_NO_BUFS);
+        SuccessOrExit(error = fragment->SetLength(aMessage.GetOffset() + sizeof(fragmentHeader) + payloadFragment));
+        header.SetPayloadLength(payloadFragment + sizeof(fragmentHeader));
+        VerifyOrExit(fragment->Write(0, sizeof(header), &header) == sizeof(header), error = OT_ERROR_NO_BUFS);
+        SuccessOrExit(error = fragment->SetOffset(aMessage.GetOffset()));
+        VerifyOrExit(fragment->Write(aMessage.GetOffset(), sizeof(fragmentHeader), &fragmentHeader) ==
+                         sizeof(fragmentHeader),
+                     error = OT_ERROR_NO_BUFS);
+        VerifyOrExit(aMessage.CopyTo(aMessage.GetOffset() + (offset << 3),
+                                     aMessage.GetOffset() + sizeof(fragmentHeader), payloadFragment,
+                                     *fragment) == (int)payloadFragment,
+                     error = OT_ERROR_NO_BUFS);
+        EnqueueDatagram(*fragment);
+
+        fragmentCnt++;
+        fragment = NULL;
+        otLogInfoIp6("Fragment %d with %d bytes sent ", fragmentCnt, payloadFragment);
+    }
+exit:
+    aMessage.Free();
+
+    if (error == OT_ERROR_NO_BUFS)
+    {
+        otLogWarnIp6("No buffer for Ip6 fragmentation");
+    }
+
+    if (error != OT_ERROR_NONE && fragment != NULL)
+    {
+        fragment->Free();
+    }
+#else
+    OT_UNUSED_VARIABLE(aMessage);
+    OT_UNUSED_VARIABLE(aIpProto);
+#endif // OPENTHREAD_ENABLE_IP6_FRAGMENTATION
+    return error;
+}
+
+otError Ip6::HandleFragment(Message &aMessage, Header &aHeader, MessageInfo &aMessageInfo)
 {
     otError        error = OT_ERROR_NONE;
     FragmentHeader fragmentHeader;
+#if OPENTHREAD_ENABLE_IP6_FRAGMENTATION
```

Comment:
Rather than inlining the `OPENTHREAD_ENABLE_IP6_FRAGMENTATION` within the method, how about having something like:

```
#if OPENTHREAD_ENABLE_IP6_FRAGMENTATION
otError Ip6::HandleFragment(...)
{
}
#else 
otError Ip6::HandleFragment(...)
{
}
#endif 
```

It might be worth having a single if-else-endif block for `OPENTHREAD_ENABLE_IP6_FRAGMENTATION` and include all the fragment-specific methods.

Reply:
Yes, it's clearer. I put the fragment-specific methods into a single if-else block.